File: C/chapitre8/Exercice5.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <sys/socket.h>
#include <netinet/in.h>
#include <pthread.h>

void *relais(void *arg) {
    int c1 = ((int*)arg)[0];
    int c2 = ((int*)arg)[1];
    char buf[256];
    while (1) {
        int n = read(c1, buf, sizeof(buf));
        if (n <= 0) break;
        buf[n] = '\0';
        char msg[300];
        sprintf(msg, "L'autre client dit: %s\n", buf);
        write(c2, msg, strlen(msg));
    }
    return NULL;
}

int main(void) {
    int sock = socket(AF_INET, SOCK_STREAM, 0);
    struct sockaddr_in addr;
    addr.sin_family = AF_INET;
    addr.sin_port = htons(5555);
    addr.sin_addr.s_addr = INADDR_ANY;
    bind(sock, (struct sockaddr *)&addr, sizeof(addr));
    listen(sock, 2);
    int c1 = accept(sock, NULL, NULL);
    int c2 = accept(sock, NULL, NULL);
    int fds1[2];
    fds1[0] = c1;
    fds1[1] = c2;
    pthread_t t1;
    pthread_create(&t1, NULL, relais, fds1);
    int fds2[2];
    fds2[0] = c2;
    fds2[1] = c1;
    pthread_t t2;
    pthread_create(&t2, NULL, relais, fds2);
    pthread_join(t1, NULL);
    pthread_join(t2, NULL);
    close(c1);
    close(c2);
    close(sock);
    return 0;
}

File: Python/chapitre8/Exercice5.py
import socket
import threading

def relais(src, dst):
    while True:
        data = src.recv(256)
        if not 
            break
        msg = data.decode().strip()
        reponse = "L'autre client dit: {}\n".format(msg)
        dst.sendall(reponse.encode())

def main():
    sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
    sock.bind(('0.0.0.0', 5555))
    sock.listen(2)
    c1, _ = sock.accept()
    c2, _ = sock.accept()
    t1 = threading.Thread(target=relais, args=(c1, c2))
    t2 = threading.Thread(target=relais, args=(c2, c1))
    t1.start()
    t2.start()
    t1.join()
    t2.join()
    c1.close()
    c2.close()
    sock.close()

if __name__ == "__main__":
    main()